framework/framework.go

```diff
@@ -1,54 +1,134 @@
 package framework
 
 import (
+	"fmt"
+	"sync"
+
+	"k8s.io/client-go/tools/cache"
+
 	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/micrologger"
 )
 
 // Config represents the configuration used to create a new operator framework.
 type Config struct {
 	// Dependencies.
-	Logger micrologger.Logger
+	Logger    micrologger.Logger
+	Resources []Resource
 }
 
 // DefaultConfig provides a default configuration to create a new operator
 // framework by best effort.
 func DefaultConfig() Config {
-	var err error
-
-	var newLogger micrologger.Logger
-	{
-		config := micrologger.DefaultConfig()
-		newLogger, err = micrologger.New(config)
-		if err != nil {
-			panic(err)
-		}
-	}
-
 	return Config{
 		// Dependencies.
-		Logger: newLogger,
+		Logger:    nil,
+		Resources: nil,
 	}
 }
 
+type Framework struct {
+	// Dependencies.
+	logger    micrologger.Logger
+	resources []Resource
+
+	// Internals.
+	mutex sync.Mutex
+}
+
 // New creates a new configured operator framework.
 func New(config Config) (*Framework, error) {
 	// Dependencies.
 	if config.Logger == nil {
 		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
 	}
+	if len(config.Resources) == 0 {
+		return nil, microerror.Maskf(invalidConfigError, "config.Resources must not be empty")
+	}
 
 	newFramework := &Framework{
 		// Dependencies.
-		logger: config.Logger,
+		logger:    config.Logger,
+		resources: config.Resources,
+
+		// Internals.
+		mutex: sync.Mutex{},
 	}
 
 	return newFramework, nil
 }
 
-type Framework struct {
-	// Dependencies.
-	logger micrologger.Logger
+// AddFunc executes the framework's ProcessCreate and ProcessUpdate functions,
+// in this order. This guarantees resource creation is always done before
+// resource updates.
+func (f *Framework) AddFunc(obj interface{}) {
+	// We lock the AddFunc/DeleteFunc to make sure only one AddFunc/DeleteFunc is
+	// executed at a time. AddFunc/DeleteFunc is not thread safe. This is
+	// important because the source of truth for the kvm-operator are Kubernetes
+	// resources. In case we would run the operator logic in parallel, we would
+	// run into race conditions.
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	f.logger.Log("debug", "executing the operator's create function")
+
+	err := ProcessCreate(obj, f.resources)
+	if err != nil {
+		f.logger.Log("error", fmt.Sprintf("%#v", err), "event", "create")
+		return
+	}
+
+	f.logger.Log("debug", "executed the operator's create function")
+
+	f.logger.Log("debug", "executing the operator's update function")
+
+	err = ProcessUpdate(obj, f.resources)
```

Comment:
As discussed the add function executes `ProcessCreate` and `ProcessUpdate`, in this order.